File: geode-core/src/main/java/org/apache/geode/cache/client/internal/function/FunctionServiceImpl.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.geode.cache.client.internal.function;

import org.apache.geode.cache.Region;
import org.apache.geode.cache.RegionService;
import org.apache.geode.cache.client.Pool;
import org.apache.geode.cache.client.PoolManager;
import org.apache.geode.cache.client.function.FunctionService;
import org.apache.geode.cache.client.internal.ProxyCache;
import org.apache.geode.cache.client.internal.ProxyRegion;
import org.apache.geode.cache.execute.Execution;
import org.apache.geode.cache.execute.Function;
import org.apache.geode.cache.execute.FunctionException;
import org.apache.geode.internal.cache.InternalRegion;
import org.apache.geode.internal.cache.execute.ServerFunctionExecutor;
import org.apache.geode.internal.cache.execute.ServerRegionFunctionExecutor;
import org.apache.geode.internal.i18n.LocalizedStrings;

import java.util.Map;

public class FunctionServiceImpl implements FunctionService {

  @Override
  public Execution onRegion(Region region) {
    if (region == null) {
      throw new FunctionException(
          LocalizedStrings.FunctionService_0_PASSED_IS_NULL.toLocalizedString("Region instance "));
    }
    ProxyCache proxyCache = null;
    String poolName = region.getAttributes().getPoolName();
    if (poolName != null) {
      Pool pool = PoolManager.find(poolName);
      if (pool.getMultiuserAuthentication()) {
        if (region instanceof ProxyRegion) {
          ProxyRegion proxyRegion = (ProxyRegion) region;
          region = proxyRegion.getRealRegion();
          proxyCache = proxyRegion.getAuthenticatedCache();
        } else {
          throw new UnsupportedOperationException();
        }
      }
    }

    if (isClientRegion(region)) {
      return new ServerRegionFunctionExecutor(region, proxyCache);
    }
    throw new FunctionException("Cannot find suitable Function Executor");
  }

  @Override
  public Execution onServer(Pool pool) {
    if (pool == null) {
      throw new FunctionException(
          LocalizedStrings.FunctionService_0_PASSED_IS_NULL.toLocalizedString("Pool instance "));
    }

    if (pool.getMultiuserAuthentication()) {
      throw new UnsupportedOperationException();
    }

    return new ServerFunctionExecutor(pool, false, null);
  }

  @Override
  public Execution onServers(Pool pool) {
    if (pool == null) {
      throw new FunctionException(
          LocalizedStrings.FunctionService_0_PASSED_IS_NULL.toLocalizedString("Pool instance "));
    }

    if (pool.getMultiuserAuthentication()) {
      throw new UnsupportedOperationException();
    }

    return new ServerFunctionExecutor(pool, true, null);
  }

  @Override
  public Execution onServer(RegionService regionService) {
    return null;
  }

  @Override
  public Execution onServers(RegionService regionService) {
    return null;
  }

  @Override
  public Function getFunction(String functionId) {
    return null;
  }

  @Override
  public void registerFunction(Function function) {

  }

  @Override
  public void unregisterFunction(String functionId) {

  }

  @Override
  public boolean isRegistered(String functionId) {
    return false;
  }

  @Override
  public Map<String, Function> getRegisteredFunctions() {
    return null;
  }

  /**
   * @return true if the method is called on a region has a {@link Pool}.
   * @since GemFire 6.0
   */
  private boolean isClientRegion(Region region) {
    return ((InternalRegion) region).hasServerProxy();

  }
}

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/function/FunctionServiceImpl.kt
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.geode.cache.client.internal.function

import org.apache.geode.cache.Region
import org.apache.geode.cache.RegionService
import org.apache.geode.cache.client.Pool
import org.apache.geode.cache.client.PoolManager
import org.apache.geode.cache.client.function.FunctionService
import org.apache.geode.cache.client.internal.InternalClientCache
import org.apache.geode.cache.client.internal.ProxyCache
import org.apache.geode.cache.client.internal.ProxyRegion
import org.apache.geode.cache.execute.Execution
import org.apache.geode.cache.execute.Function
import org.apache.geode.cache.execute.FunctionException
import org.apache.geode.cache.execute.internal.FunctionServiceManager
import org.apache.geode.internal.cache.InternalRegion
import org.apache.geode.internal.cache.execute.ServerFunctionExecutor
import org.apache.geode.internal.cache.execute.ServerRegionFunctionExecutor
import org.apache.geode.internal.i18n.LocalizedStrings

class FunctionServiceImpl : FunctionService {
    private val functionServiceManager = FunctionServiceManager()

    override fun onRegion(region: Region<*, *>?): Execution<*, *, *> {
        region?.let {
            var proxyCache: ProxyCache? = null
            var tmpRegion: Region<*, *> = it

            val poolName = it.attributes.poolName
            if (poolName != null) {
                val pool = PoolManager.find(poolName)
                if (pool.multiuserAuthentication) {
                    if (it is ProxyRegion) {
                        tmpRegion = it.realRegion
                        proxyCache = it.authenticatedCache
                    } else {
                        throw UnsupportedOperationException()
                    }
                }
            }
            if (tmpRegion is InternalRegion && isClientRegion(tmpRegion)) {
                return ServerRegionFunctionExecutor(region, proxyCache)
            }
            throw FunctionException("Cannot find suitable Function Executor")
        }
        throw FunctionException(LocalizedStrings.FunctionService_0_PASSED_IS_NULL.toLocalizedString("Region instance "))
    }

    override fun onServer(pool: Pool?): Execution<*, *, *> =
            createServerFunctionExecutorForPool(pool, false)

    override fun onServers(pool: Pool?): Execution<*, *, *> =
            createServerFunctionExecutorForPool(pool, true)

    override fun onServer(regionService: RegionService?): Execution<*, *, *> =
            createServerFunctionExecutorForRegionService(regionService, false)

    override fun onServers(regionService: RegionService?): Execution<*, *, *> =
            createServerFunctionExecutorForRegionService(regionService, true)

    private fun createServerFunctionExecutorForPool(pool: Pool?, runOnAllServers: Boolean, proxyCache: ProxyCache? = null): Execution<*, *, *> {
        pool?.let {
            if (it.multiuserAuthentication) {
                throw UnsupportedOperationException()
            }

            return ServerFunctionExecutor(pool, runOnAllServers, proxyCache, *returnServerGroupAsTypeArray(it.serverGroup))
        }
        throw FunctionException(
                LocalizedStrings.FunctionService_0_PASSED_IS_NULL.toLocalizedString("Pool instance "))
    }

    private fun createServerFunctionExecutorForRegionService(regionService: RegionService?, runOnAllServers: Boolean)
            : Execution<*, *, *> {
        regionService?.let {
            if (it is InternalClientCache) {
                if (!it.isClient) {
                    throw FunctionException("The cache was not a client cache")
                } else {
                    it.defaultPool?.let {
                        return createServerFunctionExecutorForPool(it, runOnAllServers)
                    }
                    throw FunctionException("The client cache does not have a default pool")
                }
            } else {
                val proxyCache = it as ProxyCache
                val pool = proxyCache.userAttributes.pool
                return createServerFunctionExecutorForPool(pool, runOnAllServers, proxyCache)
            }
        }
        throw FunctionException(LocalizedStrings.FunctionService_0_PASSED_IS_NULL
                .toLocalizedString("RegionService instance "))
    }

    private fun returnServerGroupAsTypeArray(serverGroup: String): Array<String> = serverGroup.split(",").toTypedArray()


    override fun getFunction(functionId: String?): Function<*> = functionServiceManager.getFunction(functionId)
    override fun registerFunction(function: Function<*>?) = functionServiceManager.registerFunction(function)
    override fun unregisterFunction(functionId: String?) = functionServiceManager.unregisterFunction(functionId)
    override fun isRegistered(functionId: String?): Boolean = functionServiceManager.isRegistered(functionId)
    override fun getRegisteredFunctions(): MutableMap<String, Function<Any>> = functionServiceManager.registeredFunctions

    /**
     * @return true if the method is called on a region has a [Pool].
     * @since GemFire 6.0
     */
    private fun isClientRegion(region: InternalRegion): Boolean = region.hasServerProxy()
}
